modules/volume/volume.go

```diff
@@ -247,37 +307,243 @@ func (m *Module) worker() {
+		"org.PulseAudio.Core1.Device", "Volume", dbus.MakeVariant([]uint32{uint32(newVol)}))
+
+	if call.Err != nil {
+		fmt.Fprintf(os.Stderr, "Could not set volume: %v\n", call.Err)
+	}
+
+	m.refresh()
+}
+
+func (m *paModule) SetMuted(muted bool) {
+	call := m.sink.Call("org.freedesktop.DBus.Properties.Set", 0,
+		"org.PulseAudio.Core1.Device", "Mute", dbus.MakeVariant(muted))
+	if call.Err != nil {
+		fmt.Fprintf(os.Stderr, "Could not set mute: %v\n", call.Err)
+	}
+	m.refresh()
+}
+
+func (m *paModule) listen(signal string, objects ...dbus.ObjectPath) error {
+	call := m.core.Call("org.PulseAudio.Core1.ListenForSignal", 0, "org.PulseAudio.Core1."+signal, objects)
+	return call.Err
+}
+
+func (m *paModule) openSink(sink dbus.ObjectPath) error {
+	m.sink = m.conn.Object("org.PulseAudio.Core1.Sink", sink)
+
+	if err := m.listen("Device.VolumeUpdated", sink); err != nil {
+		return err
+	}
+	if err := m.listen("Device.MuteUpdated", sink); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (m *paModule) openSinkByName(name string) error {
+	var path dbus.ObjectPath
+
+	err := m.core.Call("org.PulseAudio.Core1.GetSinkByName", 0, name).Store(&path)
+	if err != nil {
+		return err
+	}
+
+	return m.openSink(path)
+}
+
+func (m *paModule) openFallbackSink() error {
+	path, err := m.core.GetProperty("org.PulseAudio.Core1.FallbackSink")
+	if err != nil {
+		return err
+	}
+
+	return m.openSink(path.Value().(dbus.ObjectPath))
+}
+
+func (m *paModule) refresh() error {
+	v := Volume{}
+	v.Min = 0
+
+	max, err := m.sink.GetProperty("org.PulseAudio.Core1.Device.BaseVolume")
+	if err != nil {
+		return err
+	}
+	v.Max = int64(max.Value().(uint32))
+
+	vol, err := m.sink.GetProperty("org.PulseAudio.Core1.Device.Volume")
+	if err != nil {
+		return err
+	}
+
+	// Take the volume as the average across all channels.
+	var totalVol int64
+	channels := vol.Value().([]uint32)
+	for _, ch := range channels {
+		totalVol += int64(ch)
+	}
+	v.Vol = totalVol / int64(len(channels))
+
+	mute, err := m.sink.GetProperty("org.PulseAudio.Core1.Device.Mute")
+	if err != nil {
+		return err
+	}
+	v.Mute = mute.Value().(bool)
+
+	m.currentVolume.Set(v)
+	return nil
+}
+
+func (m *paModule) worker() error {
+	conn, err := openPulseAudio()
+	if err != nil {
+		return err
+	}
+	m.conn = conn
+	defer func() {
+		conn.Close()
+		m.conn = nil
+	}()
+
+	m.core = conn.Object("org.PulseAudio.Core1", "/org/pulseaudio/core1")
+	defer func() {
+		m.core = nil
+	}()
+
+	if m.sinkName != "" {
+		if err = m.openSinkByName(m.sinkName); err != nil {
+			return err
 		}
-		if vol, ok := v.(Volume); ok {
-			ch.Output(outputFunc(vol))
+	} else {
+		if err = m.openFallbackSink(); err != nil {
+			return err
 		}
-		select {
-		case <-sVol:
-			v, err = m.currentVolume.Get()
-		case <-sOutputFunc:
-			outputFunc = m.outputFunc.Get().(func(Volume) bar.Output)
+
+		if err = m.listen("FallbackSinkUpdated"); err != nil {
+			return err
 		}
 	}
+
+	defer func() {
```

Comment:
Can this be removed? Each Stream() call should restart worker which will overwrite previous sink anyway, and it doesn't look like sink == nil is treated specially anywhere.

Reply:
That's to handle the case where an error causes worker() to exit, thus closing the D-Bus connection. If a click happens in the interim, I don't want it to be making a call on a remote object that might not even be around anymore.

I'll add a nil check in the SetVolume and SetMuted methods for paModule.

Reply:
That works too. Eventually I'd like the bar to never send clicks to modules in error states, so that modules can be assured that if they get a click event they're in a healthy state.